easylinks/src/main/java/com/google/step/servlets/ManageServlet.java

```diff
@@ -31,13 +33,44 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
     PreparedQuery results = ServletHelper.DEFAULT_DATASTORE_SERVICE.prepare(query);
 
     response.setContentType("application/json;");
-    response.getWriter().println(ServletHelper.convertToJson(results.asList()));
+    response.getWriter().println(
+            ServletHelper.GSON.toJson(
+                  results.asList(FetchOptions.Builder.withDefaults())));
   }
 
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    long id = Long.parseLong(request.getParameter("id"));
-    Key linkEntityKey = KeyFactory.createKey("Link", id);
-    ServletHelper.DEFAULT_DATASTORE_SERVICE.delete(linkEntityKey);
+    String action = ServletHelper.getParameterWithDefault(request, "action", "");
+    switch (action)
+    {
+      case "add":
+        addLink(request, response);
+        break;
+      case "delete":
+        long id = Long.parseLong(request.getParameter("id"));
+        Key linkEntityKey = KeyFactory.createKey("Link", id);
+        ServletHelper.DEFAULT_DATASTORE_SERVICE.delete(linkEntityKey);
+        break;
+      default:
+        break;
+    }
+  }
+
+  private static void addLink(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String shortcut = ServletHelper.getParameterWithDefault(request, "shortcut", "");
+    String url = ServletHelper.getParameterWithDefault(request, "url", "");
+    if (shortcut.isEmpty() || url.isEmpty()) {
+      response.getWriter().println("Invalid Input.");
```

Comment:
Also make sure the customized shortcut cannot start with "~"?

Reply:
Done